consensus/src/quorum_store/utils.rs

```diff
@@ -0,0 +1,282 @@
+        Self {
+            mempool_tx,
+            mempool_txn_pull_timeout_ms,
+        }
+    }
+
+    pub async fn pull_internal(
+        &self,
+        max_items: u64,
+        max_bytes: u64,
+        exclude_txns: Vec<TransactionSummary>,
+    ) -> Result<Vec<SignedTransaction>, anyhow::Error> {
+        let (callback, callback_rcv) = oneshot::channel();
+        let msg = QuorumStoreRequest::GetBatchRequest(max_items, max_bytes, exclude_txns, callback);
+        self.mempool_tx
+            .clone()
+            .try_send(msg)
+            .map_err(anyhow::Error::from)?;
+        // wait for response
+        match monitor!(
+            "pull_txn",
+            timeout(
+                Duration::from_millis(self.mempool_txn_pull_timeout_ms),
+                callback_rcv
+            )
+            .await
+        ) {
+            Err(_) => Err(anyhow::anyhow!(
+                "[direct_mempool_quorum_store] did not receive GetBatchResponse on time"
+            )),
+            Ok(resp) => match resp.map_err(anyhow::Error::from)?? {
+                QuorumStoreResponse::GetBatchResponse(txns) => Ok(txns),
+                _ => Err(anyhow::anyhow!(
+                    "[direct_mempool_quorum_store] did not receive expected GetBatchResponse"
+                )),
+            },
+        }
+    }
+}
+
+// TODO: unitest
+pub struct ProofQueue {
+    digest_queue: VecDeque<(HashValue, LogicalTime)>,
+    digest_proof: HashMap<HashValue, Option<ProofOfStore>>, // None means committed
+}
+
+impl ProofQueue {
+    pub(crate) fn new() -> Self {
+        Self {
+            digest_queue: VecDeque::new(),
+            digest_proof: HashMap::new(),
+        }
+    }
+
+    pub(crate) fn push(&mut self, proof: ProofOfStore) {
+        match self.digest_proof.entry(*proof.digest()) {
+            Vacant(entry) => {
+                self.digest_queue
+                    .push_back((*proof.digest(), proof.expiration()));
+                entry.insert(Some(proof));
+            }
+            Occupied(mut entry) => {
+                if entry.get().is_some()
+                    && entry.get().as_ref().unwrap().expiration() < proof.expiration()
+                {
+                    entry.insert(Some(proof));
+                }
+            }
+        }
+    }
+
+    // gets excluded and iterates over the vector returning non excluded or expired entries.
+    pub(crate) fn pull_proofs(
+        &mut self,
+        excluded_proofs: &HashSet<HashValue>,
+        current_time: LogicalTime,
+        max_txns: u64,
+        max_bytes: u64,
+    ) -> Vec<ProofOfStore> {
+        let num_expired = self
+            .digest_queue
+            .iter()
+            .take_while(|(_, expiration_time)| *expiration_time < current_time)
+            .count();
+        for (digest, _) in self.digest_queue.drain(0..num_expired) {
+            assert_some!(self.digest_proof.remove(&digest));
+        }
```

Comment:
I think you could do this in the for-loop below. It can facilitate potentially more gc as well.

Reply:
Rati really likes take_while...